src/main/java/frc/trigon/robot/subsystems/elevator/Elevator.java

```diff
@@ -0,0 +1,130 @@
+import org.trigon.hardware.phoenix6.talonfx.TalonFXSignal;
+import org.trigon.utilities.Conversions;
+
+public class Elevator extends MotorSubsystem {
+    private final TalonFXMotor motor = ElevatorConstants.MASTER_MOTOR;
+    private final VoltageOut voltageRequest = new VoltageOut(0).withEnableFOC(ElevatorConstants.FOC_ENABLED);
+    private final MotionMagicVoltage positionRequest = new MotionMagicVoltage(0).withEnableFOC(ElevatorConstants.FOC_ENABLED);
+    private ElevatorConstants.ElevatorState targetState = ElevatorConstants.ElevatorState.REST;
+
+    public Elevator() {
+        setName("Elevator");
+    }
+
+    @Override
+    public SysIdRoutine.Config getSysIdConfig() {
+        return ElevatorConstants.SYSID_CONFIG;
+    }
+
+    @Override
+    public void setBrake(boolean brake) {
+        motor.setBrake(brake);
+    }
+
+    @Override
+    public void stop() {
+        motor.stopMotor();
+    }
+
+    @Override
+    public void updateLog(SysIdRoutineLog log) {
+        log.motor("Elevator")
+                .linearPosition(Units.Meters.of(getPositionRotations()))
+                .linearVelocity(Units.MetersPerSecond.of(motor.getSignal(TalonFXSignal.VELOCITY)))
+                .voltage(Units.Volts.of(motor.getSignal(TalonFXSignal.MOTOR_VOLTAGE)));
+    }
+
+    @Override
+    public void updateMechanism() {
+        Logger.recordOutput("Poses/Components/ElevatorFirstPose", getFirstComponentPose());
+        Logger.recordOutput("Poses/Components/ElevatorSecondPose", getSecondComponentPose());
+
+        ElevatorConstants.MECHANISM.update(
+                rotationsToMeters(getPositionRotations()),
+                rotationsToMeters(motor.getSignal(TalonFXSignal.CLOSED_LOOP_REFERENCE))
+        );
+    }
+
+    @Override
+    public void updatePeriodically() {
+        motor.update();
+    }
+
+    @Override
+    public void sysIdDrive(double targetVoltage) {
+        motor.setControl(voltageRequest.withOutput(targetVoltage));
+    }
+
+    public boolean atTargetState() {
+        final double currentToTargetStateDifference = Math.abs(targetState.targetPositionMeters - getPositionMeters());
+        return currentToTargetStateDifference < ElevatorConstants.TOLERANCE_METERS;
+    }
+
+    void setTargetState(ElevatorConstants.ElevatorState targetState) {
+        this.targetState = targetState;
+        setTargetPositionRotations(metersToRotations(targetState.targetPositionMeters));
+    }
+
+    void setTargetPositionRotations(double targetPositionRotations) {
+        motor.setControl(positionRequest.withPosition(targetPositionRotations));
+    }
+
+    private Pose3d getFirstComponentPose() {
+        final Pose3d originPoint = ElevatorConstants.FIRST_STAGE_VISUALIZATION_ORIGIN_POINT;
+
+        if (!elevatorExtendedFirstComponentLimit())
+            return calculateCurrentElevatorPoseFromOrigin(originPoint);
+        return setComponentPose(ElevatorConstants.FIRST_ELEVATOR_COMPONENT_EXTENDED_LENGTH, originPoint);
```

Comment:
apperantly it doesn't work exactly like this 
the first component is the one always moving according to the pose, and once it passed a certain threshold the second component also starts moving